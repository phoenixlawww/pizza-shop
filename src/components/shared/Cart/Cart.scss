@import "~styles/tools";

.s-cart {
  display: flex;
  flex-direction: column;
  align-items: stretch;
  justify-content: flex-start;
  padding: 10px;
  background-color: $extra-light-grey;
  border-radius: 10px;

  > .cart-empty {
    padding: 20px 0;
    text-align: center;
  }

  > .cart-items {
    padding-bottom: 10px;

    .cart-item {
      display: flex;
      flex-direction: row;
      align-items: flex-start;
      justify-content: space-between;
      column-gap: 10px;

      + .cart-item {
        margin-top: 10px;
      }

      .item-name {
        font-weight: 400;
      }

      .item-options {
        font-size: 12px;
      }

      .item-delete-btn {
        position: relative;
        display: flex;
        flex-direction: row;
        flex-shrink: 0;
        align-items: center;
        justify-content: center;
        width: 20px;
        height: 20px;
        cursor: pointer;
        background-color: $grey;
        border-radius: 50%;

        &::before,
        &::after {
          position: absolute;
          top: calc((100% - 80%) / 2);
          left: calc((100% - 2px) / 2);
          display: block;
          width: 2px;
          height: 80%;
          content: "";
          background-color: $white;
        }
        &::before {
          transform: rotateZ(-45deg);
        }
        &::after {
          transform: rotateZ(45deg);
        }
      }

      .item-detail-container {
        flex: 1;
        text-align: right;
      }

      .total-item-amount {
        flex-shrink: 0;
        min-width: 50px;
        text-align: right;
      }
    }
  }

  > .cart-total {
    display: flex;
    flex-direction: row;
    justify-content: flex-end;
    padding: 20px 0;
    padding: 5px 0 30px;
    column-gap: 10px;
    font-weight: 600;
    text-align: right;
    border-top: 1px dashed $black;

    > .cart-amount {
      min-width: 55px;
    }
  }
}
