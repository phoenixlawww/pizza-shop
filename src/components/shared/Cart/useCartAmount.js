import ShopContext from 'contexts/ShopContext'
import { useContext, useEffect, useState } from 'react'

const calculateAmount = (cartItems) =>
  cartItems.reduce((amount, cartItem) => {
    const { unitPrice, quantity } = cartItem

    return amount + unitPrice * quantity
  }, 0)

export default function useCartAmount () {
  const {
    state: { cartItems }
  } = useContext(ShopContext)

  const [cartAmount, setCartAmount] = useState(calculateAmount(cartItems))

  useEffect(() => {
    setCartAmount(calculateAmount(cartItems))
  }, [cartItems])

  return cartAmount
}
