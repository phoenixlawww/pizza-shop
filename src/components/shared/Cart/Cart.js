import PropTypes from 'prop-types'
import classnames from 'classnames'

import formatCurrency from 'utils/number/formatCurrency'

import Button, { BUTTON_VARIENTS } from 'components/shared/Button'

import './Cart.scss'
import { useContext } from 'react'
import ShopContext from 'contexts/ShopContext'
import useCartAmount from './useCartAmount'

function Cart (props) {
  const { className } = props

  const {
    state: { cartItems },
    dispatch
  } = useContext(ShopContext)

  const cartAmount = useCartAmount()

  return (
    <div className={classnames('s-cart', className)}>
      {cartItems.length === 0 ? (
        <div className="cart-empty">No Items in your basket</div>
      ) : (
        <div className="cart-items">
          {cartItems.map((item, index) => {
            const { name, unitPrice, options, quantity } = item

            return (
              <div className="cart-item" key={index}>
                <div
                  className="item-delete-btn"
                  onClick={(e) => {
                    dispatch({
                      type: 'REMOVE_PRODUCT_FROM_CART',
                      item
                    })
                  }}
                />
                <div className="item-detail-container">
                  <div className="item-name">{`${quantity} x ${name}`}</div>
                  <div className="item-options">
                    {options.map((option) => option.shortName).join(', ')}
                  </div>
                </div>
                <div className="total-item-amount">{`$${formatCurrency(
                  unitPrice * quantity
                )}`}</div>
              </div>
            )
          })}
        </div>
      )}

      <div className="cart-total">
        <div>Total</div>
        <div className="cart-amount">{`$${formatCurrency(cartAmount)}`}</div>
      </div>

      <Button
        label={'Checkout'}
        varient={BUTTON_VARIENTS.PRIMARY}
        disabled={cartItems.length === 0}
        onClick={() => alert('TODO checkout')}
      />
    </div>
  )
}

Cart.propTypes = {
  className: PropTypes.string
}

Cart.defaultProps = {}

export default Cart
