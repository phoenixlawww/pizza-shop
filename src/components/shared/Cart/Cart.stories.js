import Cart from './Cart'

import ShopContext from 'contexts/ShopContext'
import useShopReducer from 'components/view/App/useShopReducer'

import pizzaList from 'mock-data/products/pizza.json'
import pizzaOptions from 'mock-data/products/pizza/options.json'

export default {
  component: Cart,
  title: 'Shared/Cart',
  args: {},
  decorators: [
    (Story, { parameters }) => {
      const { state, dispatch } = useShopReducer({
        cartItems: parameters.cartItems
      })

      return (
        <ShopContext.Provider value={{ state, dispatch }}>
          <Story />
        </ShopContext.Provider>
      )
    },

    (Story) => {
      return (
        <div style={{ maxWidth: 300 }}>
          <Story />
        </div>
      )
    }
  ]
}

const Template = (args) => <Cart {...args} />

export const Normal = Template.bind({})
Normal.parameters = {
  cartItems: [
    {
      ...pizzaList.products[0],
      options: [
        pizzaOptions.options[0].subOptions[0],
        pizzaOptions.options[1].subOptions[0]
      ],
      quantity: 1
    },
    {
      ...pizzaList.products[1],
      options: [
        pizzaOptions.options[0].subOptions[1],
        pizzaOptions.options[0].subOptions[2],
        pizzaOptions.options[1].subOptions[1]
      ],
      quantity: 2
    }
  ]
}

export const EmptyCart = Template.bind({})
EmptyCart.parameters = {
  cartItems: []
}
