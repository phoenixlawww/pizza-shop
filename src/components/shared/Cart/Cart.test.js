import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import ShopContext from 'contexts/ShopContext'

import Cart from './Cart'

import pizzaList from 'mock-data/products/pizza.json'
import pizzaOptions from 'mock-data/products/pizza/options.json'

const customRender = (ui, { providerProps, ...renderOptions }) => {
  return render(
    <ShopContext.Provider {...providerProps}>{ui}</ShopContext.Provider>,
    renderOptions
  )
}

test('renders empty cart correctly', () => {
  const { baseElement } = render(
    <ShopContext.Provider
      value={{
        state: { cartItems: [] }
      }}
    >
      <Cart />
    </ShopContext.Provider>
  )

  expect(baseElement).toMatchSnapshot()
})

test('renders cart item correctly', () => {
  const { baseElement } = render(
    <ShopContext.Provider
      value={{
        state: {
          cartItems: [
            {
              ...pizzaList.products[0],
              options: [
                pizzaOptions.options[0].subOptions[0],
                pizzaOptions.options[1].subOptions[0]
              ],
              quantity: 1
            }
          ]
        }
      }}
    >
      <Cart />
    </ShopContext.Provider>
  )

  expect(baseElement).toMatchSnapshot()
})

test('calculate total correctly', () => {
  customRender(<Cart />, {
    providerProps: {
      value: {
        state: {
          cartItems: [
            {
              ...pizzaList.products[0],
              options: [pizzaOptions.options[1].subOptions[0]],
              quantity: 2
            },
            {
              ...pizzaList.products[1],
              options: [pizzaOptions.options[1].subOptions[0]],
              quantity: 1
            }
          ]
        }
      }
    }
  })

  expect(screen.getByText('$297')).toBeInTheDocument()
})

test('call dispatch successfully', () => {
  const dispatch = jest.fn()

  const { container } = customRender(<Cart />, {
    providerProps: {
      value: {
        state: {
          cartItems: [
            {
              ...pizzaList.products[0],
              options: [pizzaOptions.options[1].subOptions[0]],
              quantity: 1
            }
          ]
        },
        dispatch
      }
    }
  })

  userEvent.click(container.querySelector('.item-delete-btn'))
  expect(dispatch).toBeCalled()
})
