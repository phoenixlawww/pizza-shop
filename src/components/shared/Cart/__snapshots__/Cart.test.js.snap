// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders cart item correctly 1`] = `
<body>
  <div>
    <div
      class="s-cart"
    >
      <div
        class="cart-items"
      >
        <div
          class="cart-item"
        >
          <div
            class="item-delete-btn"
          />
          <div
            class="item-detail-container"
          >
            <div
              class="item-name"
            >
              1 x Pizza A
            </div>
            <div
              class="item-options"
            >
              Topping #1, Small
            </div>
          </div>
          <div
            class="total-item-amount"
          >
            $99
          </div>
        </div>
      </div>
      <div
        class="cart-total"
      >
        <div>
          Total
        </div>
        <div
          class="cart-amount"
        >
          $99
        </div>
      </div>
      <button
        class="s-button s-button-primary"
      >
        Checkout
      </button>
    </div>
  </div>
</body>
`;

exports[`renders empty cart correctly 1`] = `
<body>
  <div>
    <div
      class="s-cart"
    >
      <div
        class="cart-empty"
      >
        No Items in your basket
      </div>
      <div
        class="cart-total"
      >
        <div>
          Total
        </div>
        <div
          class="cart-amount"
        >
          $0
        </div>
      </div>
      <button
        class="s-button s-button-primary"
        disabled=""
      >
        Checkout
      </button>
    </div>
  </div>
</body>
`;
